### PR TITLE
net/wireguard: Add dashboard widget

### PR DESCRIPTION
**Pull request by x-jokay**
Adds a widget for WireGuard.

Relates #1862.

### CODE REVIEW

**Comment by x-jokay** [comment]
@mimugmail please test it and let me know if it works for you :wink:

File: net/wireguard/src/opnsense/service/conf/actions.d/actions_wireguard.conf [hunk 1]
```diff
@@ -39,3 +39,9 @@ command:/usr/local/bin/wg show all latest-handshakes
 parameters:
 type:script_output
 message:show Wireguard handshakes
+
+[showhandshake.widget]
```
Comment (AdSchellevis) [inline net/wireguard/src/opnsense/service/conf/actions.d/actions_wireguard.conf:43]: this doesn't work, showhandshake points to the action above, configd doesn't support further nesting. have you tried this locally?

On my end this crashes configd.

File: net/wireguard/src/www/widgets/wireguard.widget.php [hunk 2]
```diff
@@ -0,0 +1,45 @@
...
+ */
+
+require_once("widgets/include/wireguard.inc");
+
+$handshakes = configd_run("wireguard showhandshake widget");
+
+?>
+<table class="table table-striped table-condensed">
+    <tbody>
+        <?php foreach($handshakes as $item) {
```
Comment (AdSchellevis) [inline net/wireguard/src/www/widgets/wireguard.widget.php:36]: More of a style thing, if possible, can you replace the template loop with a construction like:
```
foreach(..):
....
endforeach;
```

It's easier to read, most of our inline php code uses this.

File: net/wireguard/src/opnsense/service/conf/actions.d/actions_wireguard.conf (hunk 1 above)
Comment (x-jokay) [reply net/wireguard/src/opnsense/service/conf/actions.d/actions_wireguard.conf:43]: I have seen something similar ([action](https://github.com/opnsense/plugins/blob/master/sysutils/smart/src/opnsense/service/conf/actions.d/actions_smart.conf#L7) & [widget](https://github.com/opnsense/plugins/blob/master/sysutils/smart/src/www/widgets/widgets/smart_status.widget.php#L34)) for the `smart widget`.

I have not tried this locally but I can rename it.

Comment (AdSchellevis) [reply net/wireguard/src/opnsense/service/conf/actions.d/actions_wireguard.conf:43]: @x-jokay which does explain this report as well https://github.com/opnsense/plugins/pull/1848 (and missing root cause analysis). The smart widget is flawed as well, leading to an unstable `configd` service.

@darkain ^^ and there's your root cause analysis.

**Comment by x-jokay** [comment]
> I do expect the next feature request will be to refresh the widget content on changes.

Absolutely, but this is what I can provide atm. To be honest, PHP is not really my strong point :wink: 
The widget will be updated on each load of the dashboard, this is imho better than nothing?

File: net/wireguard/src/opnsense/service/conf/actions.d/actions_wireguard.conf (hunk 1 above)
Comment (x-jokay) [reply net/wireguard/src/opnsense/service/conf/actions.d/actions_wireguard.conf]: :wink:

File: net/wireguard/src/www/widgets/wireguard.widget.php (hunk 2 above)
Comment (x-jokay) [reply net/wireguard/src/www/widgets/wireguard.widget.php]: Changed as requested.

**Comment by AdSchellevis** [comment]
@x-jokay no problem, let's ask @mimugmail if he accepts the feature as is so we can pull it in.  The effort is certainly appreciated, we all have to start somewhere :)

File: net/wireguard/src/opnsense/service/conf/actions.d/actions_wireguard.conf (hunk 1 above)
Comment (x-jokay) [reply net/wireguard/src/opnsense/service/conf/actions.d/actions_wireguard.conf]: Renamed to `[widget]`.

**Comment by mimugmail** [comment]
Hi,

I installed your branch but I can't find the widget in the UI?
Did you also test you patch on a clean install?

**Comment by x-jokay** [comment]
> Did you also test you patch on a clean install?

No, it's untested. I have no idea what the requirements for an OPNsense plugin are. 

This PR is only an "interpretation" from some other plugins of what might be the correct change(s) to get this widget working.

**Comment by AdSchellevis** [comment]
@x-jokay just checkout the plugins repo on an OPNsense and execute a `make upgrade` in the wireguard directory. Please make sure to test stuff before asking others for feedback, people are quite busy and you want to make sure their time is properly spend.

@mimugmail sorry about the noise, I was under the impression that it was tested, but had issues due to the `configd` bug.

**Comment by x-jokay** [comment]
@AdSchellevis like this?

```sh
user@debian:~/GitHub/core/plugins/net/wireguard$ make upgrade
Makefile:7: *** missing separator.  Stop.
```

**Comment by AdSchellevis** [comment]
on an OPNsense/HardenedBSD machine, yes, see https://github.com/opnsense/tools for information about our toolchain.

**Comment by mimugmail** [comment]
On your OPNsense installation:

```
pkg install git
git clone https://github.com/x-jokay/plugins/
cd plugins && git checkout feature/wireguard-widget
cd net/wireguard
make upgrade
```

If you do any changes in github you can easily `git pull && make upgrade`

**Comment by x-jokay** [comment]
Thx, will setup a VM for testing.

**Comment by x-jokay** [comment]
wireguard-widget-empty

wireguard-widget

**Comment by x-jokay** [comment]
@AdSchellevis & @mimugmail got it working now :+1:

**Comment by mimugmail** [comment]
@x-jokay Is it now ready for testing? :)

**Comment by x-jokay** [comment]
@mimugmail yes, can be tested now.

**Comment by mimugmail** [comment]
It works, thanks, but as soon as you have more than one column it misses a line break:

image

**Comment by x-jokay** [comment]
May be this is because of this [trim()](https://github.com/opnsense/plugins/pull/1865/files#diff-40d7158a31b1fc7ba33233bee6362c27R32).

**Comment by x-jokay** [comment]
current

It worked for me using multiple interfaces, will do some more tests.

**Comment by x-jokay** [comment]
Well, this is indeed a problem :wink:

ui-probs

@mimugmail may be you got an idea on how to get this one responsive?

**Comment by mimugmail** [comment]
I would only display first 8 chars or endpoint, I'd guess noone is interested in full output

**Comment by x-jokay** [comment]
Good idea, will do it like this.

**Comment by x-jokay** [comment]
ui-current

File: net/wireguard/pkg-descr [hunk 3]
```diff
@@ -16,12 +16,16 @@ WWW: https://www.wireguard.com/
 Changelog
 ---------
 
+1.2
+
+* Dashboard widget
```
Comment (fichtner) [inline net/wireguard/pkg-descr:21]: credit is added here if you want, usually as "(contributed by xxx)"

File: net/wireguard/src/www/widgets/widgets/wireguard.widget.php [hunk 4]
```diff
@@ -0,0 +1,82 @@
...
+ * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ * POSSIBILITY OF SUCH DAMAGE.
+ */
+
+require_once("guiconfig.inc");
+require_once("widgets/include/wireguard.inc");
+
+$data = configd_run("wireguard widget");
+
+$empty = strlen($data) <= 2;
```
Comment (fichtner) [inline net/wireguard/src/www/widgets/widgets/wireguard.widget.php:34]: maybe this works now, but it is obscure and unreliable in the future

File: net/wireguard/src/www/widgets/widgets/wireguard.widget.php [hunk 5]
```diff
@@ -0,0 +1,82 @@
...
+
+<table class="table table-striped table-condensed">
+    <thead>
+        <tr>
+            <th><?= gettext("Interface") ?></th>
+            <th><?= gettext("Endpoint") ?></th>
+            <th><?= gettext("Latest Handshake") ?></th>
+        </tr>
+    </thead>
+    <tbody>
```
Comment (fichtner) [inline net/wireguard/src/www/widgets/widgets/wireguard.widget.php:49]: in general it is preferable to always print the header and use colspan="3" to print a message for no instances with a single td

File: net/wireguard/src/www/widgets/widgets/wireguard.widget.php [hunk 6]
```diff
@@ -0,0 +1,82 @@
...
+
+<?php foreach ($handshakes as $handshake):
+    $item = explode(",", $handshake);
+
+    $epoch = $item[2];
+    $latest = "-";
+    if ($epoch > 0):
+        $dt = new DateTime("@$epoch");
+        $latest=$dt->format("Y-m-d H:i:sP");
```
Comment (fichtner) [inline net/wireguard/src/www/widgets/widgets/wireguard.widget.php:58]:
```suggestion
        $latest = $dt->format(gettext('Y-m-d H:i:sP'));
```

File: net/wireguard/src/www/widgets/widgets/wireguard.widget.php [hunk 7]
```diff
@@ -0,0 +1,82 @@
...
+    endif; ?>
+
+        <tr>
+            <td><?= gettext($item[0]) ?></td>
```
Comment (fichtner) [inline net/wireguard/src/www/widgets/widgets/wireguard.widget.php:62]: I don't think gettext is appropriate at this point

File: net/wireguard/src/www/widgets/widgets/wireguard.widget.php [hunk 8]
```diff
@@ -0,0 +1,82 @@
...
+            <td><?= gettext(substr($item[1], 0, 10)) ?>...</td>
+            <td><?= gettext($latest) ?></td>
```
Comment (fichtner) [inline net/wireguard/src/www/widgets/widgets/wireguard.widget.php:64]: same here

File: net/wireguard/src/www/widgets/widgets/wireguard.widget.php (hunk 5 above)
Comment (x-jokay) [reply net/wireguard/src/www/widgets/widgets/wireguard.widget.php:62]: I made the change but the style will be different (in comparison to the OpenVPN widget).

comparison

File: net/wireguard/src/www/widgets/widgets/wireguard.widget.php (hunk 4 above)
Comment (x-jokay) [reply net/wireguard/src/www/widgets/widgets/wireguard.widget.php]: I improved the check on empty output.
Please let me know if you know an even more reliable way.

**Comment by x-jokay** [comment]
@fichtner ready for another review.

File: net/wireguard/src/www/widgets/widgets/wireguard.widget.php (hunk 5 above)
Comment (x-jokay) [reply net/wireguard/src/www/widgets/widgets/wireguard.widget.php:62]: Or did you mean adding a `td colspan="3"` in the `tbody` section for `no instances`?

**Comment by x-jokay** [comment]
@fichtner This is how it looks now.

current

**Comment by mimugmail** [comment]
Looks good 👍

**Comment by DasSkelett** [comment]
Also tried it out on my firewall, seems to work fine!

**Comment by x-jokay** [comment]
@AdSchellevis @fichtner please let me know if anything is missing so it can be merged.

**Comment by AdSchellevis** [comment]
@x-jokay I'm very sorry, it has been extremely busy so I totally forgot about this one.

**Pull request merged**